106-auto-rebalance-on-imbalanced-clusters.md

```diff
@@ -0,0 +1,471 @@
+
+### Proposal
+
+This proposal allows the users to have their cluster balanced automatically whenever the cluster gets imbalanced due to overloaded broker, CPU usage e.t.c.
+With self-healing ability of Cruise Control, the rebalances are triggered automatically in the cluster which means that the operator wouldn't have the information about when a rebalance is happening.
+To resolve this issue, we will only make use of the ability of Cruise Control to detect the anomalies and based on the detection, we will then notify the operator to run the rebalance using an approach based on the existing auto-rebalance for scaling feature.
+We will be using the goal violation anomaly detection related classes in Cruise Control to detect imbalanced cluster.
+Doing this will provide us with the following advantages:
+* we will ensure that the operator knows what is going on in the Kafka cluster.
+* using the existing `KafkaRebalance` CR system make it easier for users to see what is happening and when, which aids in debugging.
+* ensures the operators is in charge of rebalances.
+
+### `skew` mode in Strimzi's auto-rebalancing feature
+
+The [`auto-rebalancing`](https://strimzi.io/docs/operators/latest/deploying#proc-automating-rebalances-str) feature in Strimzi allows the operator to run a rebalance automatically when a Kafka cluster is scaled up (by adding brokers) or scaled down (by removing brokers).
+
+Auto-rebalancing in Strimzi currently supports two modes:
+* add-brokers - auto-rebalancing on scale up
+* remove-brokers - auto-rebalancing on scale down
+
+To leverage the automated rebalance on imbalanced cluster, we will be introducing a new mode to the auto-rebalancing feature.
+The new mode will be called `skew`, which means that cluster imbalance was detected and rebalancing should be applied to the all the brokers.
+The mode is defined by setting the `spec.cruiseControl.autoRebalance.mode` field as `skew` and the corresponding rebalancing configuration is defined as a reference to a "template" `KafkaRebalance` custom resource, by using the `spec.cruiseControl.autoRebalance.template` field as a [LocalObjectReference](https://kubernetes.io/docs/reference/kubernetes-api/common-definitions/local-object-reference/).
+This field is optional and if not specified, the auto-rebalancing runs with the default Cruise Control configuration.
+To provide user more flexibility, they don't have to configure all the modes for auto-rebalancing. 
+They can configure auto-rebalance to enable only for their specific case i.e. setting only `skew` mode or other scaling related modes.
+Once the auto-rebalance with `skew` mode is enabled, the operator will be ready to trigger auto-rebalance whenever the cluster becomes imbalanced.
+To trigger the auto-rebalance, the operator must know that the cluster is imbalanced due to some goal violation anomaly. 
+We will create our own custom notifier named `AnomalyDetectorNotifier` to do the same.
```

Comment:
We don't know the future but we know you can configure just one notifier within Cruise Control so what about using a more generic name like `StrimziCruiseControlNotifier`, `StrimziNotifier` or similar.
Right now the notifier is doing what we are proposing but it could take other actions in the future.

Reply:
Yeah, that would make it more flexible for future changes, so +1 for naming it more generic way...